Problem with inline operators when using Dredd to mutate tint

Dredd appears to have problems mutating inline operators within tint. I have run into this issue in several different files and with multiple operators, but they all follow the example below.

### Example

Using Dredd to mutate [this](https://dawn.googlesource.com/dawn/+/e570e1c82/src/tint/lang/core/ir/transform/robustness.cc) tint file gives the following error message:

```
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/transform/robustness.cc:533:97: error: no matching member function for call to 'Constant'
                    if (!__dredd_enabled_mutation(277)) { return b.Subtract(ty.u32(), length, b.Constant(__dredd_replace_expr_unsigned_long_long_one(1_u, 273)))->Result(__dredd_replace_expr_unsigned_long_zero(0, 275)); }
                                                                                              ~~^~~~~~~~
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:331:19: note: candidate function not viable: no known conversion from 'unsigned long long' to 'const core::constant::Value *' for 1st argument
    ir::Constant* Constant(const core::constant::Value* val) {
                  ^
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:339:19: note: candidate function not viable: no known conversion from 'unsigned long long' to 'core::i32' (aka 'Number<int>') for 1st argument
    ir::Constant* Constant(core::i32 v) { return Constant(ConstantValue(v)); }
                  ^
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:344:19: note: candidate function not viable: no known conversion from 'unsigned long long' to 'core::u32' (aka 'Number<unsigned int>') for 1st argument
    ir::Constant* Constant(core::u32 v) { return Constant(ConstantValue(v)); }
                  ^
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:349:19: note: candidate function not viable: no known conversion from 'unsigned long long' to 'core::f32' (aka 'Number<float>') for 1st argument
    ir::Constant* Constant(core::f32 v) { return Constant(ConstantValue(v)); }
                  ^
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:354:19: note: candidate function not viable: no known conversion from 'unsigned long long' to 'core::f16' (aka 'Number<tint::core::detail::NumberKindF16>') for 1st argument
    ir::Constant* Constant(core::f16 v) { return Constant(ConstantValue(v)); }
                  ^
/data/dev/dawn_mutated_transform/src/tint/lang/core/ir/builder.h:360:19: note: candidate template ignored: requirement 'std::is_same_v<unsigned long long, bool>' was not satisfied [with BOOL = unsigned long long]
    ir::Constant* Constant(BOOL v) {
                  ^
```

The unmutated tint code is [here](https://dawn.googlesource.com/dawn/+/e570e1c82/src/tint/lang/core/ir/transform/robustness.cc#258), and the relevant expression is:

`b.Constant(1_u)`

Dredd mutates this to:

`b.Constant(__dredd_replace_expr_unsigned_long_long_one(1_u, 273))`

Where the function `__dredd_replace_expr_unsigned_long_long_one(unsigned long long int, int)` has return type `unsigned long long int`. However, the overloaded tint function `Constant`(declared [here](https://dawn.googlesource.com/dawn/+/e570e1c82/src/tint/lang/core/ir/builder.h#328)) only accepts a parameter of type `core::constant::Value*`, `core::i32`, `core::u32`, `core::f32`, `core::f16`, or `BOOL`.

The problem appears to be that Dredd appears is incorrectly parsing the expression `1_u`. Tint defines `_u` as an inline suffix operator [here](https://dawn.googlesource.com/dawn/+/e570e1c82/src/tint/lang/core/number.h#692):   

```
/// Literal suffix for u32 literals
inline u32 operator""_u(unsigned long long int value) {  // NOLINT
    return u32(static_cast<uint32_t>(value));
}
```

Dredd appears to not understand `_u` as an inline operator that has return type `u32`. It is correct that the value 1 in the expression `b.Constant(1_u)` has type `unsigned long long int`, however I think that the correct mutation should be _within_ the `_u` suffix like so:

`b.Constant((__dredd_replace_expr_unsigned_long_long_one(1, 273))_u)`

--- a/src/libdredd/src/mutate_visitor.cc
+++ b/src/libdredd/src/mutate_visitor.cc
@@ -144,6 +144,13 @@ bool MutateVisitor::TraverseStmt(clang::Stmt* stmt) {
     return true;
   }
 
+  // Do not mutate user defined literals.
+  // TODO(https://github.com/mc-imperial/dredd/issues/223): Consider supporting
+  // them.
+  if (llvm::dyn_cast<clang::UserDefinedLiteral>(stmt) != nullptr) {
+    return true;
+  }
+
   // Do not mutate under a constant expression, since mutation logic is
   // inherently non-constant.
   if (llvm::dyn_cast<clang::ConstantExpr>(stmt) != nullptr) {

--- a/test/single_file/user_defined_literals.cc
+++ b/test/single_file/user_defined_literals.cc
@@ -5,11 +5,7 @@ struct Number {
     
     uint32_t value = {};
     
-    // Constructors
-    Number() = default;
     explicit Number(unsigned long long int v) : value(static_cast<uint32_t>(v)) {}
-
-    operator uint32_t() const { return value; }
 
 };
 

--- a/test/single_file/user_defined_literals.cc
+++ b/test/single_file/user_defined_literals.cc
@@ -1,0 +1,27 @@
+#include <cstdint>
+
+// Number wraps integer, enforcing explicit casting
+struct Number {
+    
+    uint32_t value = {};
+    
+    // Constructors
+    Number() = default;
+    explicit Number(unsigned long long int v) : value(static_cast<uint32_t>(v)) {}
+
+    operator uint32_t() const { return value; }
+
+};
+
+using u32 = Number;
+
+inline u32 operator""_u(unsigned long long int value) {
+    return u32(static_cast<uint32_t>(value));
+}
+
+int main() {
+    
+    u32 foo = 1_u;
+
+    return 0;
+}

--- a/test/single_file/user_defined_literals.cc.expected
+++ b/test/single_file/user_defined_literals.cc.expected
@@ -1,0 +1,103 @@
+#include <cinttypes>
+#include <cstddef>
+#include <functional>
+#include <string>
+
+
+#ifdef _MSC_VER
+#define thread_local __declspec(thread)
+#elif __APPLE__
+#define thread_local __thread
+#endif
+
+static thread_local bool __dredd_some_mutation_enabled = true;
+static bool __dredd_enabled_mutation(int local_mutation_id) {
+  static thread_local bool initialized = false;
+  static thread_local uint64_t enabled_bitset[1];
+  if (!initialized) {
+    bool some_mutation_enabled = false;
+    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
+    if (dredd_environment_variable != nullptr) {
+      std::string contents(dredd_environment_variable);
+      while (true) {
+        size_t pos = contents.find(",");
+        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
+        if (!token.empty()) {
+          int value = std::stoi(token);
+          int local_value = value - 0;
+          if (local_value >= 0 && local_value < 21) {
+            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
+            some_mutation_enabled = true;
+          }
+        }
+        if (pos == std::string::npos) {
+          break;
+        }
+        contents.erase(0, pos + 1);
+      }
+    }
+    initialized = true;
+    __dredd_some_mutation_enabled = some_mutation_enabled;
+  }
+  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
+}
+
+static unsigned long long __dredd_replace_expr_unsigned_long_long_lvalue(unsigned long long& arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --(arg);
+  return arg;
+}
+
+static unsigned long long __dredd_replace_expr_unsigned_long_long(unsigned long long arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 0;
+  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 1;
+  return arg;
+}
+
+static unsigned int __dredd_replace_expr_unsigned_int(unsigned int arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 0;
+  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 1;
+  return arg;
+}
+
+static int __dredd_replace_expr_int_zero(int arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return 1;
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return -1;
+  return arg;
+}
+
+#include <cstdint>
+
+// Number wraps integer, enforcing explicit casting
+struct Number {
+    
+    uint32_t value = {};
+    
+    // Constructors
+    Number() = default;
+    explicit Number(unsigned long long int v) : value(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_long_long_lvalue(v, 0)), 2)) {}
+
+    operator uint32_t() const { if (!__dredd_enabled_mutation(10)) { return __dredd_replace_expr_unsigned_int(value, 6); } }
+
+};
+
+using u32 = Number;
+
+inline u32 operator""_u(unsigned long long int value) {
+    if (!__dredd_enabled_mutation(17)) { return u32(__dredd_replace_expr_unsigned_long_long(static_cast<uint32_t>(__dredd_replace_expr_unsigned_long_long_lvalue(value, 11)), 13)); }
+}
+
+int main() {
+    
+    u32 foo = 1_u;
+
+    if (!__dredd_enabled_mutation(20)) { return __dredd_replace_expr_int_zero(0, 18); }
+}

--- a/test/single_file/user_defined_literals.cc.expected
+++ b/test/single_file/user_defined_literals.cc.expected
@@ -25,7 +25,7 @@ static bool __dredd_enabled_mutation(int local_mutation_id) {
         if (!token.empty()) {
           int value = std::stoi(token);
           int local_value = value - 0;
-          if (local_value >= 0 && local_value < 21) {
+          if (local_value >= 0 && local_value < 16) {
             enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
             some_mutation_enabled = true;
           }
@@ -81,23 +81,19 @@ struct Number {
     
     uint32_t value = {};
     
-    // Constructors
-    Number() = default;
     explicit Number(unsigned long long int v) : value(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_long_long_lvalue(v, 0)), 2)) {}
-
-    operator uint32_t() const { if (!__dredd_enabled_mutation(10)) { return __dredd_replace_expr_unsigned_int(value, 6); } }
 
 };
 
 using u32 = Number;
 
 inline u32 operator""_u(unsigned long long int value) {
-    if (!__dredd_enabled_mutation(17)) { return u32(__dredd_replace_expr_unsigned_long_long(static_cast<uint32_t>(__dredd_replace_expr_unsigned_long_long_lvalue(value, 11)), 13)); }
+    if (!__dredd_enabled_mutation(12)) { return u32(__dredd_replace_expr_unsigned_long_long(static_cast<uint32_t>(__dredd_replace_expr_unsigned_long_long_lvalue(value, 6)), 8)); }
 }
 
 int main() {
     
     u32 foo = 1_u;
 
-    if (!__dredd_enabled_mutation(20)) { return __dredd_replace_expr_int_zero(0, 18); }
+    if (!__dredd_enabled_mutation(15)) { return __dredd_replace_expr_int_zero(0, 13); }
 }

--- a/test/single_file/user_defined_literals.cc.noopt.expected
+++ b/test/single_file/user_defined_literals.cc.noopt.expected
@@ -1,0 +1,107 @@
+#include <cinttypes>
+#include <cstddef>
+#include <functional>
+#include <string>
+
+
+#ifdef _MSC_VER
+#define thread_local __declspec(thread)
+#elif __APPLE__
+#define thread_local __thread
+#endif
+
+static thread_local bool __dredd_some_mutation_enabled = true;
+static bool __dredd_enabled_mutation(int local_mutation_id) {
+  static thread_local bool initialized = false;
+  static thread_local uint64_t enabled_bitset[1];
+  if (!initialized) {
+    bool some_mutation_enabled = false;
+    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
+    if (dredd_environment_variable != nullptr) {
+      std::string contents(dredd_environment_variable);
+      while (true) {
+        size_t pos = contents.find(",");
+        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
+        if (!token.empty()) {
+          int value = std::stoi(token);
+          int local_value = value - 0;
+          if (local_value >= 0 && local_value < 45) {
+            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
+            some_mutation_enabled = true;
+          }
+        }
+        if (pos == std::string::npos) {
+          break;
+        }
+        contents.erase(0, pos + 1);
+      }
+    }
+    initialized = true;
+    __dredd_some_mutation_enabled = some_mutation_enabled;
+  }
+  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
+}
+
+static unsigned long long __dredd_replace_expr_unsigned_long_long_lvalue(unsigned long long& arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --(arg);
+  return arg;
+}
+
+static unsigned long long __dredd_replace_expr_unsigned_long_long(unsigned long long arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 0;
+  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 1;
+  return arg;
+}
+
+static unsigned int __dredd_replace_expr_unsigned_int(unsigned int arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 0;
+  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 1;
+  return arg;
+}
+
+static int __dredd_replace_expr_int(int arg, int local_mutation_id) {
+  if (!__dredd_some_mutation_enabled) return arg;
+  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -(arg);
+  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 0;
+  if (__dredd_enabled_mutation(local_mutation_id + 4)) return 1;
+  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -1;
+  return arg;
+}
+
+#include <cstdint>
+
+// Number wraps integer, enforcing explicit casting
+struct Number {
+    
+    uint32_t value = {};
+    
+    // Constructors
+    Number() = default;
+    explicit Number(unsigned long long int v) : value(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_int(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_long_long_lvalue(v, 0), 2), 6)), 10)) {}
+
+    operator uint32_t() const { if (!__dredd_enabled_mutation(18)) { return __dredd_replace_expr_unsigned_int(value, 14); } }
+
+};
+
+using u32 = Number;
+
+inline u32 operator""_u(unsigned long long int value) {
+    if (!__dredd_enabled_mutation(37)) { return u32(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_int(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_long_long_lvalue(value, 19), 21), 25)), 29), 33)); }
+}
+
+int main() {
+    
+    u32 foo = 1_u;
+
+    if (!__dredd_enabled_mutation(44)) { return __dredd_replace_expr_int(0, 38); }
+}

--- a/test/single_file/user_defined_literals.cc.noopt.expected
+++ b/test/single_file/user_defined_literals.cc.noopt.expected
@@ -25,7 +25,7 @@ static bool __dredd_enabled_mutation(int local_mutation_id) {
         if (!token.empty()) {
           int value = std::stoi(token);
           int local_value = value - 0;
-          if (local_value >= 0 && local_value < 45) {
+          if (local_value >= 0 && local_value < 40) {
             enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
             some_mutation_enabled = true;
           }
@@ -85,23 +85,19 @@ struct Number {
     
     uint32_t value = {};
     
-    // Constructors
-    Number() = default;
     explicit Number(unsigned long long int v) : value(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_int(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_long_long_lvalue(v, 0), 2), 6)), 10)) {}
-
-    operator uint32_t() const { if (!__dredd_enabled_mutation(18)) { return __dredd_replace_expr_unsigned_int(value, 14); } }
 
 };
 
 using u32 = Number;
 
 inline u32 operator""_u(unsigned long long int value) {
-    if (!__dredd_enabled_mutation(37)) { return u32(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_int(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_long_long_lvalue(value, 19), 21), 25)), 29), 33)); }
+    if (!__dredd_enabled_mutation(32)) { return u32(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_int(static_cast<uint32_t>(__dredd_replace_expr_unsigned_int(__dredd_replace_expr_unsigned_long_long(__dredd_replace_expr_unsigned_long_long_lvalue(value, 14), 16), 20)), 24), 28)); }
 }
 
 int main() {
     
     u32 foo = 1_u;
 
-    if (!__dredd_enabled_mutation(44)) { return __dredd_replace_expr_int(0, 38); }
+    if (!__dredd_enabled_mutation(39)) { return __dredd_replace_expr_int(0, 33); }
 }

